### PR TITLE
upgrade spark version from 3.1 to 3.2, with shim

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes are proposed in this pull request?
Upgrade Spark version from 3.1 to 3.2

## How is this patch tested?
Existing tests

Note: This is a clone of https://github.com/projectglow/glow/pull/481 for us to test a shim for the private API changes in Spark 3.2.

### CODE REVIEW

File: core/src/main/shim/2.4/SparkShim.scala [hunk 1]
```diff
@@ -158,4 +160,10 @@ object SparkShim extends SparkShimBase {
       }
     }
   }
+
+  def newUnresolvedException[TreeType <: TreeNode[_]](
+      tree: TreeType,
+      function: String): Exception = {
```
Comment (williambrandler) [inline core/src/main/shim/2.4/SparkShim.scala:166]: leave out the function here?

File: core/src/main/scala/io/projectglow/sql/expressions/glueExpressions.scala [hunk 2]
```diff
@@ -21,20 +21,22 @@ import io.projectglow.sql.util.{Rewrite, RewriteAfterResolution}
...
 import org.apache.spark.sql.SQLUtils
 import org.apache.spark.sql.catalyst.InternalRow
-import org.apache.spark.sql.catalyst.analysis.{UnresolvedException, UnresolvedExtractValue}
+import org.apache.spark.sql.catalyst.analysis.UnresolvedExtractValue
 import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, CodegenFallback, ExprCode}
 import org.apache.spark.sql.catalyst.expressions.{Alias, CreateNamedStruct, ExpectsInputTypes, Expression, Generator, GenericInternalRow, GetStructField, ImplicitCastInputTypes, Literal, NamedExpression, UnaryExpression, Unevaluable}
 import org.apache.spark.sql.catalyst.util.{ArrayData, GenericArrayData}
 import org.apache.spark.sql.types._
 
+import io.projectglow.SparkShim.{newUnresolvedException, UnresolvedException}
```
Comment (williambrandler) [inline core/src/main/scala/io/projectglow/sql/expressions/glueExpressions.scala:30]: import io.projectglow.SparkShim.{newUnresolvedException}

**Review by williambrandler — COMMENTED** [review]
some comments on the errors from circleci

File: core/src/main/shim/2.4/SparkShim.scala (hunk 1 above)
Comment (jkbradley) [reply core/src/main/shim/2.4/SparkShim.scala:166]: I think we need this to compile against Spark 2.4, right?

**Comment by williambrandler** [comment]
all the tests are passing except the hail on spark 3 tests. 
Requires Hail to upgrade to Spark 3.2, let's pause until that is available

**Comment by williambrandler** [comment]
change of plan, hail will not be upgraded until spark 3.2 is widely available (on EMR, dataproc etc.), so we will press on with the release of Glow on Spark 3.2 with version 1.2.0, even though the Hail on Spark 3 tests are failing. The export hail to glow function still works on Spark 3.2, and we already recommend that users have separate clusters for hail and glow

**Pull request merged**